query/src/main/java/com/connexta/replication/query/QueryService.java

```diff
@@ -0,0 +1,223 @@
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.query;
+
+import com.connexta.replication.api.AdapterException;
+import com.connexta.replication.api.NodeAdapter;
+import com.connexta.replication.api.data.Filter;
+import com.connexta.replication.api.data.FilterIndex;
+import com.connexta.replication.api.data.Metadata;
+import com.connexta.replication.api.data.MetadataInfo;
+import com.connexta.replication.api.data.OperationType;
+import com.connexta.replication.api.data.QueryRequest;
+import com.connexta.replication.api.data.Site;
+import com.connexta.replication.api.data.TaskInfo;
+import com.connexta.replication.api.queue.QueueException;
+import com.connexta.replication.api.queue.SiteQueue;
+import com.connexta.replication.data.QueryRequestImpl;
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The <code>QueryService</code> class implements the logic for querying a specific DDF-based site
+ * for changes to be replicated or harvested.
+ */
+public class QueryService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryService.class);
+
+  private final QueryServiceTools queryServiceTools;
+
+  private final ScheduledExecutorService executor;
+
+  /** The current configuration for the site being queried by this query service. */
+  private Site site;
+
+  private SiteQueue siteQueue;
+
+  /** The adapter to communicate with the given site. */
+  private NodeAdapter adapter;
+
+  ScheduledFuture future;
+
+  /**
+   * Instantiates a new query service.
+   *
+   * @param site the configuration for the site to be queried by this query service
+   */
+  public QueryService(Site site, QueryServiceTools queryServiceTools) {
+    this(site, queryServiceTools, Executors.newSingleThreadScheduledExecutor());
+  }
+
+  public QueryService(
+      Site site, QueryServiceTools queryServiceTools, ScheduledExecutorService executor) {
+    this.queryServiceTools = queryServiceTools;
+    this.siteQueue = queryServiceTools.getQueueFor(site);
+    this.executor = executor;
+    update(site);
+  }
+
+  /**
+   * Updates the configuration for the site currently being queried by this query service.
+   *
+   * @param site the new site configuration
+   */
+  public void update(Site site) {
+    synchronized (this) {
+      this.site = site;
+      try {
+        this.adapter = queryServiceTools.getAdapterFor(site);
```

Comment:
❗️ `getAdapterFor()` can throw `IllegalArgumentException` if it is unable to find a factory for the provided site.